wallet: relax external_signer flag constraints, add musig2 test (partial)

Sjors opened the pull request:
The `external_signer` indicates that an external signer device may be called via [HWI](https://github.com/bitcoin-core/HWI) or [equivalent](https://github.com/bitcoin/bitcoin/blob/master/doc/external-signer.md#signer-api) application.

When it was initially introduced some additional constraints were placed on wallets with this flag: it had to be a descriptor wallet and watch-only. Also the flag could not be added or removed later.

The constraints aren't a problem for the main supported and documented use case of connecting a single hardware wallet and using it just like a normal single sig Bitcoin Core wallet.

But they get in the way of MuSig2 support, see https://github.com/Sjors/bitcoin/pull/91.

This pull requests drops the following constraints:

- `disable_private_keys` is no longer mandatory (but still default)
- `external_signer` flag is now mutable

Additionally it does the following:

- make the `blank` option for `createwallet` consistent with regular wallets by _not_ importing keys from the connected signer
- avoid going through `createTransaction` for external signers (otherwise the GUI breaks)

This should no noticeable effect on the default single sig use case described above.

Finally we add the beginnings of a `musig(hot key, external signer)` test. For now it just imports the descriptor and calls `walletdisplayaddress`. It takes advantage of the changes in this PR.

Sjors committed:
wallet: make watch-only optional for external signer

Before this change the external_signer flag required the wallet to be watch-only.
This precludes multisig setups in which we hold a hot key.

Remove this as a requirement, but disable private keys by default. This leaves
the typical (and only documented) use case of a single external signer unaffected.

Sjors committed:
wallet: make external_signer flag mutable

With the removal of legacy wallets and the relaxing of restrictions
in the previous commit, it's no longer a problem to toggle this flag.

Sjors committed:
wallet: don't import external keys at creation if blank

There's no need to treat external signer wallets different in this
regard. When the user sets the 'blank' flag, don't generate or
import keys.

For multisig setups that involve an external signer, it may be useful
to start from a blank wallet and manually import descriptors.

Sjors commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -379,9 +379,8 @@ static RPCHelpMan createwallet()
-        flags |= WALLET_FLAG_DISABLE_PRIVATE_KEYS;
-    }
+
+    std::optional<bool> disable_private_keys = request.params[1].isNull() ? std::nullopt : std::optional<bool>(request.params[1].get_bool());
```
That looks much nicer indeed.

Sjors commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -407,11 +406,21 @@ static RPCHelpMan createwallet()
 #endif
     }
 
+    if (disable_private_keys.has_value() && *disable_private_keys) {
```
I'll try `value_or(false)`

Sjors committed:
wallet: avoid signing via createTransaction() with external signer

External signer enabled wallets should always use the process PSBT flow.
Avoid going through CreateTransaction.

This has no effect until a later commit where WALLET_FLAG_EXTERNAL_SIGNER
no longer implies WALLET_FLAG_DISABLE_PRIVATE_KEYS. Without this change
signing with the GUI would break for external signers with private keys
enabled.

Sjors commented:
Added f8f668c355d16b8545ddc11681e6654b8f33f99a _wallet: avoid createTransaction() with signer_ to prevent breaking GUI signing for private key enabled external signer wallets (even when they don't have other keys).

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -2881,7 +2881,13 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
         // Only descriptor wallets can be created
```
In 2faf30612350fd90d9d3c44d1bb7b055addc8331 "wallet: don't import external keys at creation if blank"

Nit in commit message:
```diff
- For multisig setups than involve an external signer
+ For multisig setups that involve an external signer 
```

rkrux commented on src/qt/walletmodel.cpp:
```diff
@@ -203,7 +203,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
         int nChangePosRet = -1;
```
In 7beb338a0d4343a622236876c7d63d56bf7039e3 "wallet: avoid createTransaction() with signer"

Nit in commit message: 
```diff
- wallet: avoid createTransaction() with signer
+ wallet: avoid signing via createTransaction() with external signer
```

rkrux commented on test/functional/wallet_signer.py:
```diff
@@ -65,23 +65,22 @@ def run_test(self):
     def test_valid_signer(self):
         self.log.debug(f"-signer={self.mock_signer_path()}")
 
-        # Create new wallets for an external signer.
```
In 08f7813f536c242b7a4b65d01cfbc73601846a25 "wallet: make watch-only optional for external signer"

This comment can stay?

Sjors replied:
Brought it back.

rkrux reviewed (commented):
ACK a9734039a7a34e38145927f02b891685f96ab9e8

Agree with the intent to relax these constraints for external signer wallets.

Sjors commented on test/functional/wallet_signer.py:
```diff
@@ -65,23 +65,22 @@ def run_test(self):
+
         # Flag can't be set afterwards (could be added later for non-blank descriptor based watch-only wallets)
-        self.nodes[1].createwallet(wallet_name='not_hww', disable_private_keys=True, external_signer=False)
+        self.nodes[1].createwallet(wallet_name='not_hww', external_signer=False)
```
I think that'll become more relevant, and probably easier to test, after MuSig2 support lands.

Sjors commented:
Rebased and addressed nits.

rkrux reviewed (approved):
re-ACK 03978530ad8dc9124307d2ffc7d64c24b784be0e

```
git range-diff a973403...0397853
```

adyshimony commented on test/functional/wallet_signer.py:
```diff
@@ -66,22 +66,23 @@ def test_valid_signer(self):
-
+        not_hww.setwalletflag("external_signer", True)
+        assert_equal(not_hww.getwalletinfo()["external_signer"], True)
 
```
Check the private_keys_enabled is true by default for this case:

`assert_equal(not_hww.getwalletinfo()["private_keys_enabled"], True)`

adyshimony reviewed (commented):
ACK 7d3fc167362a

Built and ran successfully all tests.



Qt UX tested on regtest with mock signer:

- Create Wallet defaults with signer detected are correct.

- Checkbox interactions behave as expected when toggling external_signer.

- getwalletinfo for default external signer wallet shows external_signer=true, private_keys_enabled=false.

- Creating with external_signer=true and disable_private_keys manually unchecked gives external_signer=true, private_keys_enabled=true.

- setwalletflag "external_signer" set to true/false succeeds and is reflected in getwalletinfo.

naiyoma reviewed (commented):
Concept ACK

Reviewed 
3e57405bf6d1b2d75cf831537f364ec3c6092614
b990dbb504fd1b140332ca7c13f92673f74f5735
0ba76bf0a75dc7e60803dcb1e85a8492bb9397eb

and tested 500fc7a0bddee2ed7b59a8228e13637ccebd5e31 createwallet on `bitcoin-qt`

Before this pr
disable_private_keys is always checked and  grayed out, so there's no way to uncheck it
external signer always watch-only, no exceptions

<img width="705" height="432" alt="Image" src="https://github.com/user-attachments/assets/e099f7c4-9bb2-4b1d-b6dd-b08fc8e775db" />

and on cli

```
"private_keys_enabled": false,
"flags": [
    "last_hardened_xpub_cached",
    "disable_private_keys",
    "descriptor_wallet",
    "external_signer"
  ],

```
After PR, 
disabled_private_keys is checked by default, but can be unchecked

if left checked:                    
"disable_private_keys" in flags  
"private_keys_enabled": false    
same as before               


and when Unchecked
"disable_private_keys" absent
"private_keys_enabled": true
new state 

<img width="567" height="369" alt="Image" src="https://github.com/user-attachments/assets/ae80f81c-50e8-4f2b-96ff-b33a1263332b" />

and on cli

```
"descriptors": true,                                                                                                                                                                  
  "external_signer": true,                                                                                                                                                              
  "blank": false,                                                                                                                                                                       
  "birthtime": 1772971109,                                                                                                                                                              
  "flags": [                                                                                                                                                                            
    "last_hardened_xpub_cached",                                                                                                                                                        
    "descriptor_wallet",                                                                                                                                                                
    "external_signer"                                                                                                                                                                   
  ],        

```

achow101 commented:
Concept ACK

Sjors committed:
test: add MuSig2 external signer wallet test

Sjors commented:
Rebased for silent merge conflict with https://github.com/bitcoin/bitcoin/pull/34049.

I've been testing this again in light of https://github.com/Sjors/bitcoin/pull/91, which also helped refresh my memory.

I dropped the first documentation commit 3e57405bf6d1b2d75cf831537f364ec3c6092614, because #33765 is doing a more thorough job there.

Added test coverage, including a new `wallet_signer_musig2.py`. This only imports a `musig(hot wallet, external signer)` descriptor for now. Spending from it requires too many other changes that are better left for a followup.

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -3124,7 +3124,13 @@ std::shared_ptr<CWallet> CWallet::CreateNew(WalletContext& context, const std::s
+            // Fetch keys from an external signer; or
+             (wallet_creation_flags & WALLET_FLAG_EXTERNAL_SIGNER) ||
+            // Generate them, unless private keys are disabled
+            !(wallet_creation_flags & (WALLET_FLAG_DISABLE_PRIVATE_KEYS)))
```
In c05e9c13266f5004b1345cd1a967d2d57a4e9ae6 "wallet: don't import external keys at creation if blank"

Redundant parenthesis.

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 747f8ef1e0..b905b9800f 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2910,7 +2910,7 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
             // Fetch keys from an external signer; or
              (wallet_creation_flags & WALLET_FLAG_EXTERNAL_SIGNER) ||
             // Generate them, unless private keys are disabled
-            !(wallet_creation_flags & (WALLET_FLAG_DISABLE_PRIVATE_KEYS)))
+            !(wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS))
         ) {
             walletInstance->SetupDescriptorScriptPubKeyMans();
         }

```

rkrux commented on test/functional/wallet_signer.py:
```diff
@@ -87,13 +87,14 @@ def test_valid_signer(self):
         assert_equal(not_hww.getwalletinfo()["private_keys_enabled"], True)
-        assert_raises_rpc_error(-8, "Wallet flag is immutable: external_signer", not_hww.setwalletflag, "external_signer", True)
+        not_hww.setwalletflag("external_signer", True)
+        assert_equal(not_hww.getwalletinfo()["external_signer"], True)
```
In 34c447940fad77dfb291016390da80d8e4d80bb6 "wallet: make external_signer flag mutable"

Nit: the wallet name can be updated now - s/not_hww/not_hww_initially

```diff
diff --git a/test/functional/wallet_signer.py b/test/functional/wallet_signer.py
index 6a730bd66d..bb05bac0be 100755
--- a/test/functional/wallet_signer.py
+++ b/test/functional/wallet_signer.py
@@ -74,11 +74,11 @@ class WalletSignerTest(BitcoinTestFramework):
         assert_equal(hww.getwalletinfo()["private_keys_enabled"], False)
 
         # Flag can be set afterwards
-        self.nodes[1].createwallet(wallet_name='not_hww', external_signer=False)
-        not_hww = self.nodes[1].get_wallet_rpc('not_hww')
-        assert_equal(not_hww.getwalletinfo()["external_signer"], False)
-        not_hww.setwalletflag("external_signer", True)
-        assert_equal(not_hww.getwalletinfo()["external_signer"], True)
+        self.nodes[1].createwallet(wallet_name='not_hww_initially', external_signer=False)
+        not_hww_initially = self.nodes[1].get_wallet_rpc('not_hww_initially')
+        assert_equal(not_hww_initially.getwalletinfo()["external_signer"], False)
+        not_hww_initially.setwalletflag("external_signer", True)
+        assert_equal(not_hww_initially.getwalletinfo()["external_signer"], True)
 
         self.set_mock_result(self.nodes[1], '0 {"invalid json"}')
         assert_raises_rpc_error(-1, 'Unable to parse JSON',
```

Sjors replied:
Done

rkrux commented on src/wallet/rpc/spend.cpp:
```diff
@@ -173,8 +173,9 @@ UniValue SendMoney(CWallet& wallet, const CCoinControl &coin_control, std::vecto
+    // This should always try to sign, if we don't have (all) private keys, don't
+    // try to do anything here.
+    if (wallet.IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) || wallet.IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
         throw JSONRPCError(RPC_WALLET_ERROR, "Error: Private keys are disabled for this wallet");
```
In e8a54514ddf0599ede28d5b8ab3f0e2b3f27372a "wallet: avoid signing via createTransaction() with external signer"

> This has no effect until the next commit when WALLET_FLAG_EXTERNAL_SIGNER
no longer implies WALLET_FLAG_DISABLE_PRIVATE_KEYS.

After the next commit, the associated error here ("Private keys are disabled for this wallet") doesn't seem correct if it's an external signer.

Sjors replied:
Indeed. I also split this change into its own commit 996064c6db85023e0f8aa4f91277ac494821292d and added a test.

rkrux commented on test/functional/wallet_signer_musig2.py:
```diff
@@ -0,0 +1,96 @@
+class WalletSignerMuSig2Test(BitcoinTestFramework):
+    def mock_signer_path(self):
+        path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'mocks', 'signer.py')
+        return sys.executable + " " + path
```
This is duplicated with wallet_signer.py. Can't these be a part of wallet_signer.py?

Sjors replied:
The MuSig2 test is going to grow by quite a lot, at least in my current (very rough) draft: https://github.com/Sjors/bitcoin/blob/2025/06/musig2-power/test/functional/wallet_signer_musig2.py

But I can move these helpers to the test framework. Done in 0b3ddcd516301a790afcf387ffdbcb7878ca4eb5.

rkrux commented on test/functional/wallet_signer_musig2.py:
```diff
@@ -0,0 +1,96 @@
+        return sys.executable + " " + path
+
+    def set_test_params(self):
+        self.num_nodes = 2
```
2 nodes don't seem to be required, only one (self.nodes[1]) is used.

Sjors replied:
Fixed.

rkrux reviewed (commented):
Code review at 8e7ea9997293df451e98e39300fc75b83329c08e

Sjors committed:
test: move mock signer path helper to the test framework

Both rpc_signer.py and wallet_signer.py defined identical
mock_signer_path() helpers; the upcoming wallet_signer_musig2.py test
needs the same helper. Move it to BitcoinTestFramework.

Sjors commented:
Rebased (just in case) and addressed @w0xlt's feedback. In particular this adds 0b3ddcd516301a790afcf387ffdbcb7878ca4eb5 to move `mock_signer_path` to the test framework and split e8a54514ddf0599ede28d5b8ab3f0e2b3f27372a into 3359674e80e74c5c3566f3207eed3681333d1940 (GUI) and 996064c6db85023e0f8aa4f91277ac494821292d (RPC).

Sjors commented:
Rebased after #28333.

Sjors commented:
Rebased after #35424 absorbed a5eb9e13f258c001b3096a61cd6420830cd63bc2.

Sjors commented:
Rebased after #35852.

Sjors commented:
Added `wallet: upgrade to ExternalSignerScriptPubKeyMan in AddWalletDescriptor` and removed reload workaround from the test.

jeanpablojp commented on src/wallet/wallet.h:
```diff
@@ -157,7 +157,8 @@ inline constexpr uint64_t KNOWN_WALLET_FLAGS =
 inline constexpr uint64_t MUTABLE_WALLET_FLAGS =
-        WALLET_FLAG_AVOID_REUSE;
+        WALLET_FLAG_AVOID_REUSE
+    |   WALLET_FLAG_EXTERNAL_SIGNER;
```
Took a watch-only wallet that already had the device's descriptors and set the flag. Right after, `walletdisplayaddress` fails with `There is no ScriptPubKeyManager for this address` and `send` returns a PSBT without calling the device. The SPKM subclass is chosen when it's constructed, and toggling doesn't rebuild the existing ones, so after `unloadwallet`/`loadwallet`, with nothing else changed, both work. The `AddWalletDescriptor` commit already handles this on the import path. Worth doing the same on toggle, or forcing a reload?

Sjors replied:
I tried to implement having the wallet reload all descriptors, but it's actually easier to just have the RPC automatically reload the whole wallet for this flag. So I went with that.

jeanpablojp commented on src/wallet/wallet.cpp:
```diff
@@ -3810,7 +3803,12 @@ util::Result<std::reference_wrapper<DescriptorScriptPubKeyMan>> CWallet::AddWall
     } else {
-        auto new_spk_man = DescriptorScriptPubKeyMan::CreateFromImport(*this, desc, m_keypool_size, signing_provider);
+        std::unique_ptr<DescriptorScriptPubKeyMan> new_spk_man;
+        if (IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
```
Imported a descriptor with the flag on, turned the flag off, imported another one, and without reloading I can't spend from either. Log excerpt:

```
  -- flag is OFF, wallet NOT reloaded
    spend the utxo imported while flag was ON    RAISED  External signer failed to sign (-25)
    spend the utxo imported while flag was OFF   RAISED  External signer failed to sign (-25)
  -- same wallet, after unload+load, flag still OFF
    spend the utxo imported while flag was ON    OK
    spend the utxo imported while flag was OFF   OK
```

The second one is what got me, since it was imported after I turned the flag off. `FillPSBT` returns on the first error, so the leftover external signer SPKM aborts the whole fill. On the previous head, `2e39d5c032`, all four succeeded.

Sjors replied:
That should be fixed now, thanks to the reload.

jeanpablojp commented on src/wallet/wallet.h:
```diff
@@ -157,7 +157,8 @@ inline constexpr uint64_t KNOWN_WALLET_FLAGS =
     |   WALLET_FLAG_EXTERNAL_SIGNER;
 
 inline constexpr uint64_t MUTABLE_WALLET_FLAGS =
```
On a build with `-DENABLE_EXTERNAL_SIGNER=OFF` I set the flag on an ordinary wallet and it no longer loads on that binary, not even after a restart, with `External signer wallet being loaded without external signer support compiled`. Since unsetting the flag requires loading the wallet, the simplest way back is a binary built with support. On the merge base the same call returns `Wallet flag is immutable: external_signer`. There's no `ENABLE_EXTERNAL_SIGNER` guard here or in `setwalletflag`, would one make sense?

Sjors replied:
Added a guard for completeness.

jeanpablojp reviewed (commented):
Concept ACK

Built and ran the wallet and signer tests.

PraneethGunas commented:
Concept ACK. This greatly helps unblock using external signers and hot keys in the same wallet for a multisig setup

PraneethGunas commented on src/wallet/wallet.cpp:
```diff
@@ -3807,7 +3807,12 @@ util::Result<std::reference_wrapper<DescriptorScriptPubKeyMan>> CWallet::AddWall
     } else {
-        auto new_spk_man = DescriptorScriptPubKeyMan::CreateFromImport(*this, desc, m_keypool_size, signing_provider);
+        std::unique_ptr<DescriptorScriptPubKeyMan> new_spk_man;
+        if (IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
```
Now that 034375fc9c allows private keys in a signer wallet, a descriptor holding a hot key also gets an ExternalSignerScriptPubKeyMan here, and at LoadDescriptorScriptPubKeyMan. Its FillPSBT() goes straight to the device when sign=true and never falls back.

Tested on a Ledger Flex, regtest. Spending a hot key UTXO gives:

  error code: -25, External signer failed to sign
  debug.log: Signer fingerprint 42036eee does not match any of the inputs:

Should SPKM selection be per descriptor, based on whether the wallet holds the key, rather than per wallet?

Sjors replied:
Don't do that :-)

I added 5a7c4309da24d9dd3ff5a8f7edc3d316cd9355c0 to have the wallet first sign with its own keys, and only try the external signer if the result is unsigned.

However we shouldn't encourage users to put hot and cold descriptors into the same wallet, so this isn't documented.

Sjors replied:
While working on this, I realized our external signer mock can't properly test these more complicated scenarios. That's only going to get worse with MuSig2, so I opened #36114 to rewrite it.

PraneethGunas replied:
Confirmed fixed on a Ledger Flex, thanks. If mixing hot and cold in one wallet
isn't encouraged, what's the intended layout for a multisig where one cosigner is
an external signer and another is a hot key, one wallet per key?

Sjors replied:
Ideally there's a single multisig descriptor for which the wallet has one key, e.g. `musig2(core_xprv, ledger_xpub)`. So you start with a blank wallet, with an `usused()` descriptor, then call `gethdkey` to export the xprv for `m/87'/0'/0'`, get the device xpub, construct the descriptor and then import it. Having to copy an xprv around is not ideal, so https://github.com/Sjors/bitcoin/pull/91 contains a commit that lets you use the xpub instead, and the wallet figures it out on import.

Sjors commented:
- added handling for hot descriptors, see https://github.com/bitcoin/bitcoin/pull/33112#discussion_r3876752265 
- dropped `test: add MuSig2 external signer wallet test` to trim the scope a bit.

PraneethGunas commented:
ACK [f4dd056](https://github.com/bitcoin/bitcoin/pull/33112/commits/f4dd05644f3eada3656f619f71ab3b94fe08ec3d), [5a7c430](https://github.com/bitcoin/bitcoin/pull/33112/commits/5a7c4309da24d9dd3ff5a8f7edc3d316cd9355c0)
